subgraphs/osmosis/src/modules/Swap.ts

```diff
@@ -0,0 +1,404 @@
+    inputTokenBalances[tokenInIndex] = inputTokenBalances[tokenInIndex].plus(
+      tokenInAmount
+    );
+    inputTokenAmounts[tokenInIndex] = tokenInAmount;
+  }
+
+  const tokenOutIndex = inputTokens.indexOf(tokenOutDenom);
+  if (tokenOutIndex >= 0) {
+    inputTokenBalances[tokenOutIndex] = inputTokenBalances[tokenOutIndex].minus(
+      tokenOutAmount
+    );
+    if (inputTokenBalances[tokenOutIndex] <= constants.BIGINT_ZERO) {
+      return;
+    }
+    inputTokenAmounts[tokenOutIndex] = tokenOutAmount;
+  }
+  // log.warning("swap() inputTokenAmounts[0] {} inputTokenAmounts[1] {}", [
+  //   inputTokenAmounts[0].toString(),
+  //   inputTokenAmounts[1].toString(),
+  // ]);
+
+  // if (
+  //   liquidityPool.inputTokenBalances[1] != constants.BIGINT_ZERO &&
+  //   inputTokenBalances[1] != constants.BIGINT_ZERO &&
+  //   inputTokenAmounts[1] != constants.BIGINT_ZERO
+  // ) {
+  //   // if (
+  //   //   inputTokenBalances[0] >= constants.BIGINT_ZERO &&
+  //   //   inputTokenBalances[1] >= constants.BIGINT_ZERO
+  //   // ) {
+
+  //   let price = liquidityPool.inputTokenBalances[0].divDecimal(
+  //     inputTokenBalances[1].toBigDecimal()
+  //   );
+
+  //   if (tokenInIndex == 0) {
+  //     price = inputTokenBalances[0].divDecimal(
+  //       liquidityPool.inputTokenBalances[1].toBigDecimal()
+  //     );
+  //   }
+  //   log.warning(
+  //     "swap for pool {} inputTokenBalances before [0] {} [1] {} ratio {} *******after [0] {} [1] {} ratio {} ****** tokenAmount [0] {} [1] {} ratio {} price {}",
+  //     [
+  //       liquidityPool.id.toString(),
+  //       liquidityPool.inputTokenBalances[0].toString(),
+  //       liquidityPool.inputTokenBalances[1].toString(),
+  //       liquidityPool.inputTokenBalances[0]
+  //         .divDecimal(liquidityPool.inputTokenBalances[1].toBigDecimal())
+  //         .toString(),
+  //       inputTokenBalances[0].toString(),
+  //       inputTokenBalances[1].toString(),
+  //       inputTokenBalances[0]
+  //         .divDecimal(inputTokenBalances[1].toBigDecimal())
+  //         .toString(),
+  //       inputTokenAmounts[0].toString(),
+  //       inputTokenAmounts[1].toString(),
+  //       inputTokenAmounts[0]
+  //         .divDecimal(inputTokenAmounts[1].toBigDecimal())
+  //         .toString(),
+  //       price.toString(),
+  //     ]
+  //   );
+  // }
+
+  liquidityPool.inputTokenBalances = inputTokenBalances;
+  liquidityPool._inputTokenAmounts = inputTokenAmounts;
+  liquidityPool.save();
+
+  let amountInUSD = constants.BIGDECIMAL_ZERO;
+  let amountOutUSD = constants.BIGDECIMAL_ZERO;
+  let volumeUSD = constants.BIGDECIMAL_ZERO;
+  const prevTVL = liquidityPool.totalValueLockedUSD;
+
+  const hasPriceData = utils.updatePoolTVL(liquidityPool, block);
+  if (hasPriceData) {
+    const tokenIn = getOrCreateToken(tokenInDenom);
+    const tokenOut = getOrCreateToken(tokenOutDenom);
+
+    updatePrice(tokenIn, tokenInAmount, tokenOut, tokenOutAmount, block.header);
```

Comment:
Something seems off to me about the price updating here. The token price is saved inside updatePrice(), but it does not return the updated token, so when you check the below conditions, it will not have the updated price. Or do I misunderstand?

Reply:
tokenIn and tokenOut are both Token object, so when lastPriceUSD entity for tokenIn or tokenOut object changes and saves in updatePrice() function, it will be reflected in the main function after calling updatePrice() function, I think.

Reply:
I think if you were to save it in the updatePrice() you would still have to reload the token to get the prices. Here is a test I ran: 

<img width="484" alt="Screen Shot 2022-12-22 at 10 17 14 AM" src="https://user-images.githubusercontent.com/56660047/209165357-1aaa1ead-93b7-45f4-b575-5c60ea9d65ab.png">

<img width="468" alt="Screen Shot 2022-12-22 at 10 17 31 AM" src="https://user-images.githubusercontent.com/56660047/209165422-a29fa472-7a48-49c1-a565-e6ab763d3868.png">

The value of the token_price variable does not update after being changed in the function. If you were to save the new price in an object, though, and then reload the object from the store after the function, then the object would be loaded with the new price I believe.

Reply:
The token_price in the above case is a simple type while tokenIn and tokenOut in my case are token object. For class object, the changes in object's entity in a function will be reflected in the calling function, as shown as below screenshot. However, it doesn't hurt to save the object in the function and then load the object again in the caller. Made the changes at 21703cbe62252f89cb45c85d27afd200b45fc0ca

![screen0](https://user-images.githubusercontent.com/28422812/209310331-39480c75-c24f-4811-87e4-7dd50fbeeefa.jpg)
![screen1](https://user-images.githubusercontent.com/28422812/209310366-8a377970-dd45-4953-a14a-a320f7a550bb.jpg)